### PR TITLE
feat(action): github action to automatically add label "release-x.y.z" when PRs get merged

### PR DESCRIPTION
**Pull request by gauthierpetetin**
## Explanation

This a new Github action that automatically adds a `release-x.y.z` label to PRs (and their linked issues) when they get merged.

The next release version number is defined with the following logic:
- Take the highest version number, based on the list of release branches (branches with Version-vx.y.z format for Extension or release/x.y.z format for Mobile)
- Take the highest version number, based on the list of tags (tags with vx.y.z format)
- Take the version number indicated in package.json
- Use the highest version number amongst those values, and do a minor version bump to deduce the next version number

For edge cases (e.g. major version bumps, etc.), where the value can not be deduced automatically, the next release version number can be set manually by defining FORCE_NEXT_SEMVER_VERSION variable in section "Secrets and variables">"Actions">"Variables">"New repository variable" in the settings of this repo.

The `release-x.y.z` label helps identify which issues and PRs will be included in the next release.

## Prerequisites

Create a personal access token with "repo" permissions.
- Go to the repository's settings and add the token as a secret named `RELEASE_LABEL_TOKEN `.
- The personal access token is necessary for creating, adding, and retrieving labels in the linked issues' repositories.

## Manual Testing Steps

1. In a first repository, create an issue.
2. In a second repository, create another issue.
3. In one of the two repositories, create a PR.
4. Link both issues to the PR.
5. Merge the PR.

## Acceptance Criteria

Upon merging the PR, the PR and both issues should have the `release-x.y.z` label.
This process should work for both public and private repositories.

## Screenshots/Recordings

https://recordit.co/aJ2gkTF8lk

## Issue

None

## Checklist

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [x] Any added code is fully documented

### CODE REVIEW

**Comment by wachunei** [comment]
Cool! I enjoyed the recording :D Is this restricted to PRs that are merged to main? Maybe we want to do that, I've seen feature branches that collect tasks:
```
main <-- PR -- feature/branch <-- PR-- feature/branch-task-1
main <-- PR -- feature/branch <-- PR-- feature/branch-task-2
```

and more often we fix bugs in a release branch:
```
release/x.y.z <-- PR -- fix/a-fix-branch
```

Maybe issues related to these edge cases would have manual labeling? What do you think?

**Comment by gauthierpetetin** [comment]
Hi @wachunei ,
Thanks for the review, I just saw tour message and indeed good catch, we want it to apply only for PRs that are merged to main. I need to change it.

**Review by tommasini — APPROVED** [review]
LGTM! The graphql it's cool ahah

**Review by tommasini — CHANGES_REQUESTED** [review]
AAAH it's supposed to be just a comment I will request changes to invalidate the approve, when this is done I take a look again 🙏

File: .github/scripts/add_release_label_to_pr_and_linked_issues.ts [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    // NEXT_RELEASE_CUT_NUMBER needs to be updated every time a new release is cut.
+    // Example value: 6.5
+    core.setFailed('NEXT_RELEASE_CUT_NUMBER not found');
+    process.exit(1);
+  }
+
+  // Release label indicates the next release cut number
+  // Example release label: "release-6.5"
+  const releaseLabelName = `release-${nextReleaseCutNumber}`;
+  const releaseLabelColor = "000000";
```
[the remark this answers is not in the document]
Comment (gauthierpetetin) [reply .github/scripts/add_release_label_to_pr_and_linked_issues.ts:44]: Thanks for the review and suggestion @tommasini . Why not, I wonder what people will prefer, I'll ask on [Slack](https://consensys.slack.com/archives/C047WAG4RFT/p1684478296218209).

[the remark this answers is not in the document]
Comment (gauthierpetetin) [reply .github/scripts/add_release_label_to_pr_and_linked_issues.ts]: I'm resolving this thread after results from poll on [Slack](https://consensys.slack.com/archives/C047WAG4RFT/p1684478296218209).

**Comment by Gudahtt** [comment]
I can review this PR as well once the conflicts have been addressed

**Comment by gauthierpetetin** [comment]
Hi @Gudahtt , yes that would be great, I resolved conflicts.

- It shall be very close to the [Extension PR](https://github.com/MetaMask/metamask-extension/pull/19061) you already approved.
- Here, I had to add this [commit](https://github.com/MetaMask/metamask-mobile/pull/6352/commits/bfcf881fdf20c542c7149e62c83f06a5d071000d) to pass the CI.

File: .github/workflows/add-release-label.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+        with:
+            fetch-depth: 0 # This is needed to checkout all branches
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: '16'
```
[the remark this answers is not in the document]
Comment (gauthierpetetin) [reply .github/workflows/add-release-label.yml]: This is now fixed with this [commit](https://github.com/MetaMask/metamask-mobile/pull/6352/commits/38872b1613267eaf9aa07afeaff02062953e8b31).

File: .github/workflows/add-release-label.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+      - name: Install dependencies
+        run: yarn --immutable
+
+      - name: Set execute permissions for script
+        run: chmod +x ./scripts/get-next-semver-version.sh
```
[the remark this answers is not in the document]
Comment (gauthierpetetin) [reply .github/workflows/add-release-label.yml]: This is now fixed with this [commit](https://github.com/MetaMask/metamask-mobile/pull/6352/commits/9c905e19754334cef6f69654bccee435f8471577).

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Comment by gauthierpetetin** [comment]
@Gudahtt , I see you already did the same changes on Extension side, thanks!

**Review by NicolasMassart — COMMENTED** [review]
LGTM only if we also create an issue reminding us to extract all this process to a standalone action once we will have validated that it works fine. Right now, it overloads the app repo with deployment CI that will require devs to load unnecessary dev dependencies when they should be handled at the action level. That's what GH actions are for.

**Comment by sethkfman** [comment]
@SocketSecurity ignore-all

All packages are dev dependencies and do not impact production code

**Review by sethkfman — APPROVED** [review]
LGTM

**Comment by sethkfman** [comment]
@gauthierpetetin PR looks good. If @tommasini doesn't complete his review we can move it forward tomorrow.

**Comment by gauthierpetetin** [comment]
Thank you all for the reviews!

@NicolasMassart, good point, I created this [issue](https://github.com/MetaMask/mobile-planning/issues/1057) in our backlog.

**Review by tommasini — APPROVED** [review]
LGTM! Awesome!!

**Comment by Gudahtt** [comment]
`RELEASE_LABEL_TOKEN` has been added as a repository secret

**Pull request merged**